routes/user.js

```diff
@@ -2,100 +2,122 @@
+}
 
-router.post("/add", (req, res) => {
+router.post("/add", async (req, res) => {
   // TODO: use a hash of the email address instead of a random string.
-  const state = crypto.randomBytes(40).toString("hex");
-  req.session.state = state;
+  const state = Subscribers.generateToken();
   const email = req.body.email;
-  req.session.email = email;
-  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}`;
-
-  EmailUtils.sendEmail(email, "Firefox Breach Alert",
-    `Visit this link to subscribe: ${url}`)
-    .then(info => {
-      // TODO: set a timer to clear this after an arbitrary timeout period.
-      gUnverifiedEmails.set(state, email);
-      res.json({
-        email,
-        info: "sent verification link",
-        // Send the would-be link back to the client in dummy mode.
-        // eslint-disable-next-line no-process-env
-        link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
-      });
-    })
-    .catch(error => {
-      res.json({ email, info: "failed to send verification link", error });
-      throw error;
+
+  if (!email)  {
+    res.status(400).json({
+      error_code: ResponseCodes.EmailNotProvided,
+      info: "Request did not include email address.",
     });
+    return;
+  }
+
+  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}&email=${email}`;
+
+  const { error } = await Subscribers.addTempUser(email, state);
+
+  if (error) {
+    resDatabaseError(res, error);
+    return;
+  }
+
+  try {
+    await EmailUtils.sendEmail(email, "Firefox Breach Alert",
+      `Visit this link to subscribe: ${url}`);
+
+    res.status(202).json({
+      info: "Sent verification link",
+      // Send the would-be link back to the client in dummy mode.
+      // eslint-disable-next-line no-process-env
+      link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
+    });
+  } catch (e) {
+    console.log(e);
+    res.status(500).json({
+      error_code: ResponseCodes.InternalError,
+      info: "SMTP error.",
+    });
+  }
 });
 
-router.get("/verify", (req, res) => {
-  const email = gUnverifiedEmails.get(req.query.state);
-  if (email) {
-    gUnverifiedEmails.delete(req.query.state);
-    Subscribers.addUser(email).then(() => {
-      res.json({ email, info: `Successfully added ${email}`});
-    }).catch((error) => {
-      res.json({ info: "Failed to add email to database.", error });
+router.get("/verify", async (req, res) => {
+  // eslint-disable-next-line prefer-const
+  let { email, token, error } = await Subscribers.getTempUser(req.query.email);
+
+  if (error) {
+    resDatabaseError(res, error);
+    return;
+  }
+
+  if (!email) {
+    res.status(400).json({
+      error_code: ResponseCodes.EmailNotFound,
+      info: "Email not marked for verification.",
     });
     return;
   }
-  // TODO: Needs better error message.
-  res.json({ info: "Who are you?" });
+
+  // Invalidate the entry even if the token doesn't match.
+  ({ error } = await Subscribers.deleteTempUser(email));
```

Comment:
Does this open a DOS attack against email addresses? I.e., an attacker could repeatedly send requests to `/verify?email=victim@example.com` so their temp user record would be deleted before they could ever verify it?

Reply:
Do we plan on having some cron script which will auto-delete temp users after X hours/days?

Reply:
I would prefer to have that, yes.

Reply:
Nice :D